tensorflow_quantum/core/ops/noise/tfq_noisy_samples.cc

```diff
@@ -0,0 +1,348 @@
+
+#include <stdlib.h>
+
+#include <string>
+
+#include "../qsim/lib/channel.h"
+#include "../qsim/lib/channels_cirq.h"
+#include "../qsim/lib/circuit.h"
+#include "../qsim/lib/circuit_noisy.h"
+#include "../qsim/lib/fuser_mqubit.h"
+#include "../qsim/lib/gate_appl.h"
+#include "../qsim/lib/gates_cirq.h"
+#include "../qsim/lib/io.h"
+#include "../qsim/lib/qtrajectory.h"
+#include "../qsim/lib/seqfor.h"
+#include "../qsim/lib/simmux.h"
+#include "cirq/google/api/v2/program.pb.h"
+#include "tensorflow/core/framework/op_kernel.h"
+#include "tensorflow/core/framework/shape_inference.h"
+#include "tensorflow/core/framework/tensor_shape.h"
+#include "tensorflow/core/lib/core/error_codes.pb.h"
+#include "tensorflow/core/lib/core/status.h"
+#include "tensorflow/core/lib/core/threadpool.h"
+#include "tensorflow_quantum/core/ops/parse_context.h"
+#include "tensorflow_quantum/core/src/circuit_parser_qsim.h"
+#include "tensorflow_quantum/core/src/util_qsim.h"
+
+namespace tfq {
+
+using ::cirq::google::api::v2::Program;
+using ::tensorflow::Status;
+
+typedef qsim::Cirq::GateCirq<float> QsimGate;
+typedef qsim::Circuit<QsimGate> QsimCircuit;
+typedef qsim::NoisyCircuit<QsimGate> NoisyQsimCircuit;
+
+class TfqNoisySamplesOp : public tensorflow::OpKernel {
+ public:
+  explicit TfqNoisySamplesOp(tensorflow::OpKernelConstruction* context)
+      : OpKernel(context) {}
+
+  void Compute(tensorflow::OpKernelContext* context) override {
+    // TODO (mbbrough): add more dimension checks for other inputs here.
+    DCHECK_EQ(4, context->num_inputs());
+
+    // Parse to Program Proto and num_qubits.
+    std::vector<Program> programs;
+    std::vector<int> num_qubits;
+    OP_REQUIRES_OK(context,
+                   GetProgramsAndNumQubits(context, &programs, &num_qubits));
+
+    // Parse symbol maps for parameter resolution in the circuits.
+    std::vector<SymbolMap> maps;
+    OP_REQUIRES_OK(context, GetSymbolMaps(context, &maps));
+    OP_REQUIRES(
+        context, maps.size() == programs.size(),
+        tensorflow::errors::InvalidArgument(absl::StrCat(
+            "Number of circuits and values do not match. Got ", programs.size(),
+            " circuits and ", maps.size(), " values.")));
+
+    int num_samples = 0;
+    OP_REQUIRES_OK(context, GetIndividualSample(context, &num_samples));
+
+    // Construct qsim circuits.
+    std::vector<NoisyQsimCircuit> qsim_circuits(programs.size(),
+                                                NoisyQsimCircuit());
+
+    Status parse_status = Status::OK();
+    auto p_lock = tensorflow::mutex();
+    auto construct_f = [&](int start, int end) {
+      for (int i = start; i < end; i++) {
+        auto r = NoisyQsimCircuitFromProgram(
+            programs[i], maps[i], num_qubits[i], true, &qsim_circuits[i]);
+        if (!r.ok()) {
+          p_lock.lock();
+          parse_status = r;
+          p_lock.unlock();
+          return;
+        }
```

Comment:
`NESTED_FN_STATUS_SYNC` after merging https://github.com/tensorflow/quantum/pull/519

Reply:
Since it's merged, please apply this at https://github.com/tensorflow/quantum/pull/519